### PR TITLE
feat(hogql): WITH support & macros

### PR DESCRIPTION
**Pull request by mariusandra**
## Changes

This PR implements "macros". Think "[lisp macros](https://lisp-journey.gitlab.io/blog/common-lisp-macros-by-example-tutorial/)" or "[C++ preprocessor macros](https://www.scaler.com/topics/cpp-macro/)". 

A macro is a piece of code that will just replace other code. In this case you define macros with the `WITH` keyword. This means we will not send anything in `WITH` to ClickHouse, but inline these expressions into the final generated query ourselves. Unlike ClickHouse's WITH, we also do support recursion here.

image

<details><summary>Generated ClickHouse Query</summary>
<p>

```sql
   SELECT final.email,
          final.screen_width
     FROM (
             SELECT customers.email,
                    customer_orders.screen_width
               FROM (
                       SELECT any (id),
                              replaceRegexpAll (JSONExtractRaw (properties, % (hogql_val_0) s), '^"|"$', '') AS email
                         FROM person
                        WHERE
                          and (
                              equals (team_id, 1),
                              isNotNull (replaceRegexpAll (JSONExtractRaw (properties, % (hogql_val_1) s),'^"|"$',''))
                              )
                     GROUP BY replaceRegexpAll (
                              JSONExtractRaw (properties, % (hogql_val_2) s),'^"|"$','')
                    ) AS customers
          LEFT JOIN (
                       SELECT avg(
                              toInt64OrNull (
                              replaceRegexpAll (
                              JSONExtractRaw (events.properties, % (hogql_val_3) s),'^"|"$',''))
                              ) AS screen_width,
                              replaceRegexpAll (
                              JSONExtractRaw (events__pdi__person.properties, % (hogql_val_4) s),'^"|"$',''
                              ) AS email
                         FROM events
                        INNER JOIN (
                                 SELECT argMax (person_distinct_id2.person_id, version) AS person_id,
                                        distinct_id
                                   FROM person_distinct_id2
                                  WHERE equals (team_id, 1)
                               GROUP BY distinct_id
                                 HAVING equals (argMax (is_deleted, version), 0)
                              ) AS events__pdi ON equals (events.distinct_id, events__pdi.distinct_id)
                        INNER JOIN (
                                 SELECT argMax (person.properties, version) AS properties,
                                        id
                                   FROM person
                                  WHERE equals (team_id, 1)
                               GROUP BY id
                                 HAVING equals (argMax (is_deleted, version), 0)
                              ) AS events__pdi__person ON equals (events__pdi.person_id, events__pdi__person.id)
                        WHERE
                          and (
                              equals (team_id, 1),
                              equals (event, % (hogql_val_5) s)
                              )
                     GROUP BY event,
                              replaceRegexpAll (
                              JSONExtractRaw (events__pdi__person.properties, % (hogql_val_6) s),'^"|"$','')
                    ) AS customer_orders ON equals (customer_orders.email, customers.email)
          ) AS final
    LIMIT 100
```
</p>
</details>

## How did you test this code?

Wrote some tests, will write more. 

## Future plans with macros:

- Save macros as database table columns. This has been called "calculated properties" in the past.
- Save select queries as virtual tables you can select from. Passing where clauses down into the lowest query is possible.
- Become dbt: Using temporal, save periodic snapshots of those virtual tables. We'd invisibly swap out the backends in queries.
- Macros with arguments (save lambdas on tables).

### CODE REVIEW

File: posthog/hogql/grammar/HogQLParser.py
Comment (yakkomajuri) [inline posthog/hogql/grammar/HogQLParser.py:1]: are we meant to just take these changes for granted?

**Review by fuziontech — APPROVED** [review]
I do love lisp and macros
🚢 it

**Comment by Twixes** [comment]
BTW I wonder why avoid ClickHouse `WITH`?

**Comment by mariusandra** [comment]
This PR is 3 weeks old, so I find it hard to remember exactly why I did things in this order. I think the main point was recursion. ClickHouse's docs say "Recursion is prevented by hiding the current level CTEs from the WITH expression.". That means you can't define an expression in WITH and have the next thing you define already use that expression. I think this is practically useful, see the query on the screenshot.

🤔

**Pull request merged**